### DISCUSSION TITLE
Typescript in dynamic slots

### DISCUSSION THREAD

**Discussion by eduardojustos** [quasarframework/quasar#14415] [General - Components / Directives / etc] [answered]
I was creating a component that wraps a QInput, which passes everything down to it, but `vue-tsc` is complaining of the following: 

```
 TS7053: Element implicitly has an 'any' type because expression of type 'string | number' can't be used to index type '{ default: unknown; prepend: unknown; append: unknown; before: unknown; after: unknown; label: unknown; error: unknown; hint: unknown; counter: unknown; loading: unknown; }'.
  No index signature with a parameter of type 'string' was found on type '{ default: unknown; prepend: unknown; append: unknown; before: unknown; after: unknown; label: unknown; error: unknown; hint: unknown; counter: unknown; loading: unknown; }'.

14         v-slot:[slot]="scope"
                  ~~~~~~
```

Component code:

```
<q-input
      borderless
      v-bind="$attrs"
      :model-value="props.modelValue"
    >
      <template
        v-for="(_, slot) in $slots"
        v-slot:[slot]="scope"
      >
        <slot
          :name="slot"
          v-bind="scope || {}"
        />
      </template>
    </q-input>
```

I was wondering if there is a workaround on this

**Comment by yusufkandemir** [comment, marked as answer]
You can do this:
```html
<template>
  <q-input
    borderless
    v-bind="$attrs"
    :model-value="props.modelValue"
  >
    <template
      v-for="(_, slot) in ($slots as Readonly<QInputSlots>)"
      v-slot:[slot]="scope"
    >
      <slot
        :name="slot"
        v-bind="scope || {}"
      />
    </template>
  </q-input>
</template>

<script setup lang="ts">
import type { QInputSlots } from 'quasar';

// ...
</script>

<script>
export default {
  inheritAttrs: false,
}
</script>
```

However, you won't get auto-completion on slot names nor get type info on template scopes. I made a comment here to see if this is expected Volar behavior: https://github.com/johnsoncodehk/volar/issues/953#issuecomment-1250029543

**Reply by cohlar** [reply]
Hi, thank you for this response, that's exactly what I was looking for... except it doesn't fix the TS issue for me.
I use VS Code 1.76.2 with TypeScript Vue Plugin (Volar) 1.2.0.

![image](https://user-images.githubusercontent.com/57757008/226887511-d9f356af-4f61-4316-99bf-a50f943b3084.png)

**Reply by yusufkandemir** [reply]
@cohlar you are using `QTable` but passing `QInputSlots` to it, use `QTableSlots` instead.

**Reply by cohlar** [reply]
Haha of course, thank you!

**Reply by cohlar** [reply]
Oh well I'm getting a bunch of other TS issues with your solution, because (I assume) table has dynamic slots which use the columns names.

Instead I went with the below - sharing in case anyone stumbles across this, and in case you know of a more elegant one:

```html
<template v-for="slot in slots" v-slot:[slot]="scope">
   <slot :name="slot" v-bind="scope || {}"></slot>
</template>
```

```
const slots = Object.keys(useSlots()) as `body-cell-${string}`[];
```

**Reply by yusufkandemir** [reply]
Works fine.
```html
<template>
  <q-table>
    <template
      v-for="(_, slot) in ($slots as Readonly<QTableSlots>)"
      #[slot]="scope"
    >
      <slot :name="slot" v-bind="scope || {}" />
    </template>
  </q-table>
</template>

<script lang="ts" setup>
import { QTableSlots } from 'quasar';
</script>
```

![image](https://user-images.githubusercontent.com/6266078/226905600-cc196774-9cc7-4a69-b2ab-6d857e4c9dac.png)

**Reply by cohlar** [reply]
I made a [Stackblitz](https://stackblitz.com/edit/vitejs-vite-wjotiq) to show the error. It doesn't show in the Stackblitz IDE, but if you open the project in your local IDE (for me VS Code), you should see the below error:

![image](https://user-images.githubusercontent.com/57757008/226970424-aeb086f4-d02a-42d0-afe4-7230d20f5496.png)

This error disappears if I use `const slots = Object.keys(useSlots()) as `body-cell-${string}`[];`.

**Comment by eduardojustos** [comment]
Thank you so much for your answer.
Unfortunately, trying this I got an issue in Webstorm: 

<img width="649" alt="image" src="https://user-images.githubusercontent.com/96435579/191009846-dc5edfa0-13e9-4d42-8331-8e8dab118662.png">

**Reply by yusufkandemir** [reply]
Please reply to answers instead of creating new "answer" entries.

You didn't mention you were using WebStorm.
See this issue for `as`: https://youtrack.jetbrains.com/issue/WEB-55049/TypeScript-syntax-as-type-cast-keyword-in-Vue-3-template
See this issue for TS in templates in general: https://youtrack.jetbrains.com/issue/WEB-54347/Use-TypeScript-language-service-in-Vue-template-expressions

Also, you didn't say so but I assume it solved the `vue-tsc` issue as you originally asked. If so, please mark it as the answer.

As an extra piece of information, `vue-tsc` is based on Volar. So, if you use VS Code + Volar, you can get the same feedback you are getting from `vue-tsc` right in your IDE.

**Reply by eduardojustos** [reply]
Sorry for that.
It fixes the issue, btw. Marked as the correct answer. Thank you so much for your help and for taking some time to provide me a better understanding of the issue

**Comment by junaidjahan** [comment]
I am still getting this issue
<img width="988" alt="Screenshot 2024-01-10 at 12 52 18 AM" src="https://github.com/quasarframework/quasar/assets/70884217/693a6bde-541a-4bd3-9ee5-9bce48f0f81e">

**Reply by tomvdk** [reply]
We solved it by using a spread operator on scope. 
```html
<template
        v-for="(_, slot) in ($slots as Readonly<QCheckboxSlots>)"
        #[slot]="...scope">
        <slot
          :name="slot"
          v-bind="(scope as any)"></slot>
</template>
```

However be aware when upgrading to prettier 3.x, it will remove the parenthesis around the $slots and scope cast. Found no way disabling this yet. Only this suggestion works: [#3821 (vuejs/language-tools)](https://github.com/vuejs/language-tools/issues/3821)

**Reply by AyazRahman** [reply]
@tomvdk While that solved the IDE error, when we run vue-tsc we get the following error:
![image](https://github.com/quasarframework/quasar/assets/6667401/e6fcae42-39f0-4845-bb59-cad6fb436a2a)
Any idea how to solve this?
